sources/appendix-value-extraction.asciidoc

```diff
@@ -0,0 +1,437 @@
+
+Allowing to put `@Valid` on type parameters allows to express the subject of cascaded validation more specifically: `Map<@Valid AddressType, @Valid Address> addresses` would describe that both, map keys and values, should be validated.
+
+=== Requirements
+
+1. Constraints can be applied to type parameters (of bean properties, but also method parameters etc.): `Property<@Min(1) Integer> value`
+1. Individual type parameters can be marked for cascaded validation: `Map<@Valid AddressType, Address> addresses`
+1. How values are retrieved must be customizable by means of a value extractor SPI, with a set of defined default extractors
+1. Sometimes constraints should apply to a value wrapped by a container type, but there is no type parameter to put the constraints to. JavaFX's `Property` hierarchy is the most prominent example: `@Email StringProperty emailProperty`. In such case the value should be implicitly extracted from the container if unambiguously doable.
+1. OPEN: Support for validation of nested collections: `List<Map<String, @NotNull String>> addressesByType`.
+1. Allow to configure type-level constraints and cascades via XML mappings
+1. Expose meta-data on type-level constraints and cascades in the constraint metadata API
+
+=== Non-requirements
+
+1. Constraints on type uses in local variables, invocations etc: `@NotNull String name = "Emmanuel";`, `new @NonEmpty @Readonly List<String>(myNonEmptyStringSet)`
+1. constraints on type uses in type definitions:
+
+[source,java]
+----
+class CustList extends List<@NotNull Customer> {
+}
+----
+
+=== Solution
+
+Value retrieval for cascaded validation and validation of type-level constraints is done via the `ValueExtractor` SPI:
+
+[source,java]
+----
+package javax.validation.valueretrieval;
+
+public interface ValueExtractor<T> {
+
+	void extractValues(T originalValue, ValueReceiver receiver);
+
+	interface ValueReceiver {
+
+		void value(Object object, String nodeName);
+
+		void iterableValue(Object object, String nodeName);
+
+		void indexedValue(Object object, String nodeName, int i);
+
+		void keyedValue(Object object, String nodeName, Object key);
+	}
+}
+----
+
+An extractor is tied to one specific type parameter of the type from which it extracts values. The `ExtractedValue` annotation is used to mark that type parameter. 
+
+As an example, this is how the implementation of the list extractor may look like:
+
+[source,java]
+----
+class ListValueExtractor implements ValueExtractor<List<@ExtractedValue ?>> {
+
+	@Override
+	public void extractValues(List<?> originalValue, ValueReceiver receiver) {
+		int i = 0;
```

Comment:
There was a good reason for it, I'm sure ;)

Let me try @Sanne-style: It'll perform much better if the incoming list is a `LinkedList` ;)